firestore/bulkwriter.go

```diff
@@ -0,0 +1,332 @@
+	}
+
+	return &BulkWriterJob{
+		op: bw.write(w[0]),
+	}, nil
+}
+
+// Update adds a document update write to the queue of writes to send.
+func (bw *BulkWriter) Update(doc *DocumentRef, updates []Update, preconds ...Precondition) (*BulkWriterJob, error) {
+	err := bw.checkWriteConditions(doc)
+	if err != nil {
+		return nil, err
+	}
+
+	w, err := doc.newUpdatePathWrites(updates, preconds)
+	if err != nil {
+		return nil, fmt.Errorf("firestore: cannot update doc %v", doc.ID)
+	}
+	return &BulkWriterJob{
+		op: bw.write(w[0]),
+	}, nil
+}
+
+// checkConditions determines whether this write attempt is valid. It returns
+// an error if either the BulkWriter has already been closed or if it
+// receives a nil document reference.
+func (bw *BulkWriter) checkWriteConditions(doc *DocumentRef) error {
+	if !bw.isOpen {
+		return errors.New("firestore: BulkWriter has been closed")
+	}
+
+	if doc == nil {
+		return errors.New("firestore: nil document contents")
+	}
+
+	_, havePath := bw.docUpdatePaths[doc.shortPath]
+	if havePath {
+		return fmt.Errorf("firestore: bulkwriter: received duplicate write for path: %v", doc.shortPath)
+	}
+
+	bw.docUpdatePaths[doc.shortPath] = true
+
+	return nil
+}
+
+// write packages up write requests into bulkWriterJob objects.
+func (bw *BulkWriter) write(w *pb.Write) bulkWriterOperation {
+	r := make(chan *pb.WriteResult, 1)
+	e := make(chan error, 1)
+
+	j := bulkWriterOperation{
+		result: r,
+		write:  w,
+		err:    e,
+		size:   0, // ignore operation size constraints; can't be inferred at compile time
+	}
+
+	err := bw.bundler.Add(j, j.size)
+	if err != nil {
+		j.err <- err
+		j.result <- nil
+	}
+	return j
+}
+
+// processResults checks for errors returned from send() and packages up the
+// results as WriteResult objects
+func processResults(o bulkWriterOperation) (*WriteResult, error) {
+	wpb := <-o.result
+	err := <-o.err
+
+	if err != nil {
+		return nil, err
+	}
+
+	wr, err := writeResultFromProto(wpb)
+	if err != nil {
+		return nil, err
+	}
+	return wr, nil
+}
+
+// send transmits writes to the service and matches response results to job channels.
+func (bw *BulkWriter) send(i interface{}) {
+	bwj := i.([]bulkWriterOperation)
+
+	var ws []*pb.Write
+	for _, w := range bwj {
+		ws = append(ws, w.write)
+	}
+
+	bwr := pb.BatchWriteRequest{
+		Database: bw.database,
+		Writes:   ws,
+		Labels:   map[string]string{},
+	}
+
+	select {
+	case <-bw.ctx.Done():
+		break
+	default:
+		resp, err := bw.vc.BatchWrite(bw.ctx, &bwr)
+		if err != nil {
+			// Do we need to be selective about what kind of errors we send?
+			for _, j := range bwj {
+				j.result <- nil
+				j.err <- err
+			}
+			return
+		}
+		// Iterate over the response. Match successful requests with unsuccessful
+		// requests.
+		for i, res := range resp.WriteResults {
+			s := resp.Status[i]
+
+			c := s.GetCode()
+
+			if c != 0 { // Should we do an explicit check against rpc.Code enum?
+				bw.retryBundler.Add(bwj[i], bwj[i].size)
```

Comment:
This seems like it things could just keep getting sent to the retry infinitly, is that true?